core/src/test/java/feign/FeignTest.java

```diff
@@ -905,6 +905,44 @@ public void beanQueryMapEncoderWithEmptyParams() throws Exception {
         .hasQueryParams("/");
   }
 
+  @Test
+  public void sharedPathParameterIsIncluded() throws Exception {
+    TestInterface api = new TestInterfaceBuilder().queryMapEndcoder(new BeanQueryMapEncoder())
+        .target("http://localhost:" + server.getPort());
+    api.configure("sharedValue");
```

Comment:
Feign `Target` objects are immutable once created.  This breaks that encapsulation and allows for the underlying `MethodMetadata` objects to be modified after the target has been built.  

In addition, there are thread-safety concerns with the mutability.  Feign Target's must be thread safe.  To keep this mechanism, the `ConfigurationMethod` must be synchronized.  

I recommend looking into alternative ways that don't require invoking a method on the Target, but using the `Contract` parsing on the `Builder` directly.  These approaches are already thread-safe, since Target construction must be done all at once.

Reply:
Current implementation does indeed sacrifice immutability for the sake of clean, explicit configuration methods. This isn't great.

The best possible solution (imo) I can think of would be for the configuration methods to generate a new target, i.e.

```java
interface ClientApi {
  @Shared
  ClientApi withEnv(@Param("env") String env);
}

ClientApi api = Feign.builder()
  .target(ClientApi.class, "http://host") // creates Target object A
  .withEnv("prod");                       // creates Target object B, object A is unchanged
```

The above is thread-safe and doesn't mutate the initially created target.

This does make the requirements around configuration methods a bit more strict - their return type always needs to be the interface that declares them - but this isn't a particularly hard rule to enforce or follow.